adafruit_connection_manager.py

```diff
@@ -0,0 +1,362 @@
+        ) -> None:
+            """Connect to a remote socket at the provided (host, port) address. The conntype
+            kwarg optionally may indicate SSL or not, depending on the underlying interface.
+            """
+
+    class SupportsRecvWithFlags(Protocol):
+        """Describes a type that posseses a socket recv() method supporting the flags kwarg."""
+
+        def recv(self, bufsize: int = ..., flags: int = ...) -> bytes:
+            """Receive data from the socket. The return value is a bytes object representing
+            the data received. The maximum amount of data to be received at once is specified
+            by bufsize. The meaning of the optional flags kwarg is implementation-specific.
+            """
+
+    class SupportsRecvInto(Protocol):
+        """Describes a type that possesses a socket recv_into() method."""
+
+        def recv_into(
+            self, buffer: bytearray, nbytes: int = ..., flags: int = ...
+        ) -> int:
+            """Receive up to nbytes bytes from the socket, storing the data into the provided
+            buffer. If nbytes is not specified (or 0), receive up to the size available in the
+            given buffer. The meaning of the optional flags kwarg is implementation-specific.
+            Returns the number of bytes received."""
+
+    class CircuitPythonSocketType(
+        CommonCircuitPythonSocketType,
+        SupportsRecvInto,
+        SupportsRecvWithFlags,
+        Protocol,
+    ):
+        """Describes the structure every modern CircuitPython socket type must have."""
+
+    class StandardPythonSocketType(
+        CommonSocketType, SupportsRecvInto, SupportsRecvWithFlags, Protocol
+    ):
+        """Describes the structure every standard Python socket type must have."""
+
+        def connect(self, address: Union[Tuple[Any, ...], str, bytes]) -> None:
+            """Connect to a remote socket at the provided address."""
+
+    SocketType = Union[
+        CircuitPythonSocketType,
+        StandardPythonSocketType,
+    ]
+
+    SocketpoolModuleType = ModuleType
+
+    class InterfaceType(Protocol):
+        """Describes the structure every interface type must have."""
+
+        @property
+        def TLS_MODE(self) -> int:  # pylint: disable=invalid-name
+            """Constant representing that a socket's connection mode is TLS."""
+
+    SSLContextType = Union[SSLContext, "_FakeSSLContext"]
+
+
+# custon exceptions
+
+
+class SocketConnectMemoryError(OSError):
+    """ConnectionManager Exception class for an MemoryError when connecting a socket."""
+
+
+class SocketConnectOSError(OSError):
+    """ConnectionManager Exception class for an OSError when connecting a socket."""
+
+
+class SocketGetOSError(OSError):
+    """ConnectionManager Exception class for an OSError when getting a socket."""
+
+
+class SocketGetRuntimeError(RuntimeError):
+    """ConnectionManager Exception class for an RuntimeError when getting a socket."""
```

Comment:
These added so `Adafruit_CircuitPython_MiniMQTT` can handle retries internally.

Reply:
These seem weird to me. Shouldn't the outer code know the context already?

Reply:
This is from [MiniMQTT._get_connect_socket](https://github.com/adafruit/Adafruit_CircuitPython_MiniMQTT/blob/a05b19f522fae5232f7d92edcd88160c4de4ed28/adafruit_minimqtt/adafruit_minimqtt.py#L304) where different exceptions are flagged as a `TemporaryError` so it knows if it's retriable or not

Reply:
These are removed